### DISCUSSION TITLE
Use Cognito provided JWT; no client secret is available

### DISCUSSION THREAD

**Discussion by revmischa** [nextauthjs/next-auth#5145] [Help] [answered]
If I set a random `clientSecret` on my Cognito provider and do the flow I get the error:
```
https://next-auth.js.org/errors#callback_oauth_error invalid_client OPError: invalid_client
    at processResponse (/Users/cyber/dev/sst-prisma/node_modules/.pnpm/openid-client@5.1.8/node_modules/openid-client/lib/helpers/process_response.js:38:13)
    at Client.grant (/Users/cyber/dev/sst-prisma/node_modules/.pnpm/openid-client@5.1.8/node_modules/openid-client/lib/client.js:1325:22)
    at async Client.callback (/Users/cyber/dev/sst-prisma/node_modules/.pnpm/openid-client@5.1.8/node_modules/openid-client/lib/client.js:474:24)
    at async oAuthCallback (/Users/cyber/dev/sst-prisma/node_modules/.pnpm/next-auth@4.10.3_biqbaboplfbrettd7655fr4n2y/node_modules/next-auth/core/lib/oauth/callback.js:114:16)
```

Since cognito is doing the token issuing for us, why do we need a secret for next-auth? We can trust that cognito has issued the token.

My cognito client does not have a client secret either:
<img width="229" alt="Screen Shot 2022-08-11 at 4 25 13 PM" src="https://user-images.githubusercontent.com/245131/184258679-e7a02ed2-9366-4c1a-bf2b-a7064f653a02.png">

I believe our Next app should just consume the tokens issued by cognito and not do any signing of its own. Like Amplify does.

Setting:
```ts
      clientSecret: '',
      client: {
        token_endpoint_auth_method: 'none',
      },
```

Almost works. The first time I auth with cognito I get `nonce mismatch, expected undefined, got: ABC123BAA.....` but after that I can log in and out fine.

_Originally posted by @revmischa in https://github.com/nextauthjs/next-auth/issues/2708#issuecomment-1212579498_

**Comment by revmischa** [comment]
Related: I also need the cognito access token for authorizing to other AWS services: https://github.com/nextauthjs/next-auth/issues/4954#issuecomment-1212615931

**Comment by balazsorban44** [comment, marked as answer]
See: 

https://github.com/nextauthjs/next-auth/issues/4706#issuecomment-1153901585
https://github.com/nextauthjs/next-auth/issues/4707
https://next-auth.js.org/configuration/callbacks#jwt-callback

**Comment by revmischa** [comment]
Even if I set the provider secret to `''` I still get errors about `NEXTAUTH_SECRET` not being set.
```
[next-auth][warn][NO_SECRET] 
https://next-auth.js.org/warnings#no_secretm
```

Should next-auth simply be verifying the JWT that Cognito is issuing? Can I tell next-auth not to generate its own tokens?

**Reply by EgidioCaprino** [reply]
@revmischa did you find a solution? I'm having the same problem. I need the `idToken` to authenticate the user on AppSync

**Reply by EgidioCaprino** [reply]
I ended up saving the `id_token` in the JWT and passing that to the client `session`

```typescript
const authOptions = {
  providers: [
    CognitoProvider({
      clientId: process.env.COGNITO_CLIENT_ID as string,
      clientSecret: process.env.COGNITO_CLIENT_SECRET as string,
      issuer: process.env.COGNITO_ISSUER,
    }),
  ],
  callbacks: {
    async jwt({ account, token }) {
      if (account) {
        // eslint-disable-next-line no-param-reassign
        token.idToken = account.id_token;
      }
      return token;
    },
    async session({ session, token }) {
      // eslint-disable-next-line no-param-reassign
      session.idToken = token.idToken;
      return session;
    },
  },
};
```
